packages/react/src/ReactStartTransition.js

```diff
@@ -9,12 +9,32 @@
 
 import ReactCurrentBatchConfig from './ReactCurrentBatchConfig';
 
+/**
+ *  If a certain number of fibers are updated inside startTransition then we suspect
+ *  that a user-space defined subscription is being wrapped by startTransition.
+ *  This is no good, they should instead use built-in APIs for concurrent mode to work correctly.
+ */
+
 export function startTransition(scope: () => void) {
   const prevTransition = ReactCurrentBatchConfig.transition;
   ReactCurrentBatchConfig.transition = 1;
   try {
     scope();
   } finally {
     ReactCurrentBatchConfig.transition = prevTransition;
+    if (__DEV__ && ReactCurrentBatchConfig._updatedFibers) {
```

Comment:
Since we need to add this same block to the hooks version of startTransition, what we can do is export a function called something like `warnIfSubscriptionDetected`. Then call it in the finally block here:

https://github.com/facebook/react/blob/d4d1dc085d1332fdd5025c292adcb5b39811e5b5/packages/react-reconciler/src/ReactFiberHooks.new.js#L1861-L1864